tools-local/unix/tool-helper.sh

```diff
@@ -0,0 +1,272 @@
+fail()
+{
+    repoRoot="$1"
+    exit_if_arg_empty "repository-root" "$repoRoot"
+    log_message "$@"
+    shift
+    echo "$@"
+    exit 1
+}
+
+# Invokes the override extension script if available, else invokes the base implementation.
+invoke_extension()
+{
+    # Displays the usage for invoke_extension function.
+    invoke_extension_usage()
+    {
+        echo "usage: invoke_extension <script-name> <repository-root> <tool-name> <override-scripts-folder-path> [args ...]"
+        echo "script-name                       Name of the extension script."
+        echo "repository-root                   Path to repository root."
+        echo "tool-name                         Name of the tool."
+        echo "override-scripts-folder-path      If a path is specified then, search and acquire scripts from the specified folder will be invoked. Otherwise, search will use the default search and acquire scripts located within the repository."
+        echo "args                              Any additional arguments that will be passed to the extension script."
+        echo ""
+        echo "Checks if the specified tool has its own implementation of the search or acquire script. If so, invokes the corresponding script. Otherwise, invokes the base implementation."
+        echo ""
+        echo "Example #1"
+        echo "invoke_extension search-tool.sh \"/Users/dotnet/corefx\" cmake \"\" \"\""
+        echo "Searches for CMake, not necessarily the declared version, using the default scripts located within the repository."
+        echo ""
+        echo "Example #2"
+        echo "invoke_extension acquire-tool.sh \"/Users/dotnet/corefx\" cmake \"\""
+        echo "Acquires the declared version of CMake, using the default scripts located within the repository."
+        echo ""
+        echo "Example #3"
+        echo "invoke_extension search-tool.sh \"/Users/dotnet/corefx\" cmake \"/Users/dotnet/MyCustomScripts\" strict"
+        echo "Searches for the declared version of CMake using the search scripts located under \"/Users/dotnet/MyCustomScripts\"."
+        echo ""
+        echo "Example #4"
+        echo "invoke_extension get-version.sh \"/Users/dotnet/corefx\" cmake \"\" \"\" \"/Users/dotnet/corefx/Tools/download/cmake/bin/cmake\" "
+        echo "Gets the version number of CMake executable located at /Users/dotnet/corefx/Tools/download/cmake/bin/cmake\"."
+        echo ""
+    }
+
+    if [ $# -lt 4 ]; then
+        invoke_extension_usage
+        exit 1
+    fi
+
+    extensionScriptName="$1"
+    repoRoot="$2"
+    toolName="$3"
+    overrideScriptsFolderPath="$4"
+    defaultScriptsFolderPath="$(get_default_scripts_folder $repoRoot)"
+
+    exit_if_arg_empty "script-name" "$extensionScriptName"
+    exit_if_arg_empty "repository-root" "$repoRoot"
+    exit_if_arg_empty "tool-name" "$toolName"
+
+    if [ ! -z "$overrideScriptsFolderPath" ] && [ ! -d "$overrideScriptsFolderPath" ]; then
```

Comment:
I think `&&` is wrong here: the message says `or`. Also, I don't think `! -z` is needed.